### PR TITLE
[Documentation][Spec Decode] Add documentation about lossless guarantees in Speculative Decoding in vLLM

### PR DESCRIPTION
**Pull request by sroy745**
Add documentation about lossless guarantees in Speculative Decoding in vLLM. The documentation documents the finding in this issue here #7627 

@cadedaniel

### CODE REVIEW

File: docs/source/models/spec_decode.rst [hunk 1]
```diff
@@ -161,6 +161,46 @@ A variety of speculative models of this type are available on HF hub:
...
+    - **Greedy Sampling Equality**: Confirms that greedy sampling with speculative decoding matches greedy sampling
+      without it. This verifies that vLLM's speculative decoding framework, when integrated with the vLLM forward pass and the vLLM rejection sampler, 
+      provides a lossless guarantee.  Almost all of the tests in `this directory <https://github.com/vllm-project/vllm/tree/b67ae00cdbbe1a58ffc8ff170f0c8d79044a684a/tests/spec_decode/e2e>`_
+      verify this property using `this assertion implementation <https://github.com/vllm-project/vllm/blob/b67ae00cdbbe1a58ffc8ff170f0c8d79044a684a/tests/spec_decode/e2e/conftest.py#L291>`_
+
+3. **vLLM Logprob Stability**
+   - vLLM currently does not guarantee stable log probabilities (logprobs) across different batch sizes, which might 
+   cause small variations in output probabilities. 
+   This issue may stem from non-deterministic behaviors in batched operations or numerical instability in Torch operations. 
+   as explained in the `Numerical Accuracy section <https://pytorch.org/docs/stable/notes/numerical_accuracy.html#batched-computations-or-slice-computations>`_
```
Comment (njhill) [inline docs/source/models/spec_decode.rst:190]: @sroy745 this isn't spec decoding specific, it applies generally when concurrent requests are batched differently. I guess would be good to have a dedicated section explaining that too...

File: docs/source/models/spec_decode.rst [hunk 2]
```diff
@@ -161,6 +161,46 @@ A variety of speculative models of this type are available on HF hub:
...
+speculative decoding, breaking down the guarantees into three key areas:
+
+1. **Theoretical Losslessness**
+   - Speculative decoding sampling is theoretically lossless up to the precision limits of hardware numerics. Floating-point errors might 
+   cause slight variations in output distributions, as discussed 
+   in `Accelerating Large Language Model Decoding with Speculative Sampling <https://arxiv.org/pdf/2302.01318>`_
+
+2. **Algorithmic Losslessness**
+   - vLLM’s implementation of speculative decoding is algorithmically validated to be lossless when the
+   temperature parameter (`temp`) is set to 0. Key tests include:
```
Comment (cadedaniel) [inline docs/source/models/spec_decode.rst:176]: the rejection sampler convergence tests also handle the case where temperature is nonzero, and/or other sampling parameters are applied.

File: docs/source/models/spec_decode.rst (hunk 1 above)
Comment (sroy745) [reply docs/source/models/spec_decode.rst]: I added a section for this in serving/faq.rst (I could not find any other generic place to add it. As you mentioned it is not specific to spec decode so thought of adding it to serving faqs).  I added a link to it in this subsection. I am not sure if this is what you meant. PTAL and let me know.

File: docs/source/models/spec_decode.rst (hunk 2 above)
Comment (sroy745) [reply docs/source/models/spec_decode.rst]: Done. Removed mention of  temperature = 0 in the comment.

**Review by sroy745 — COMMENTED** [review]
Thanks for the review. Addressed your comments.

**Review by mgoin — APPROVED** [review]

**Review by njhill — COMMENTED** [review]
Thanks @sroy745, looks great, sorry for the late review!

**Pull request merged**